File: MonglePet/Runtime/MotionScheduler.swift
nonisolated struct ScheduledMotion: Equatable, Sendable {
    let sequenceID: String
    let stepIndex: Int
    let requestedMotionID: String
    let motion: PetMotion
    let playbackSpeed: Double
    let isInteraction: Bool

    var usesIdleFallback: Bool {
        requestedMotionID != motion.id
    }
}

nonisolated enum MotionSchedulerStatus: Equatable, Sendable {
    case stopped
    case playing(ScheduledMotion)
    case unavailable
}

nonisolated struct MotionScheduler: Sendable {
    private struct Cursor: Equatable, Sendable {
        let sequence: BehaviorSequence
        var stepIndex: Int
        var remainingDuration: Duration
        var isComplete: Bool

        init(sequence: BehaviorSequence) {
            self.sequence = sequence
            stepIndex = 0
            remainingDuration = sequence.steps[0].duration
            isComplete = false
        }
    }

    private let petDefinition: PetDefinition
    private let interactionCooldown: Duration
    private var baseCursor: Cursor?
    private var pendingSequence: BehaviorSequence?
    private var interactionCursor: Cursor?
    private var cooldownRemaining: Duration = .zero
    private(set) var isPaused = false

    init(
        petDefinition: PetDefinition,
        interactionCooldown: Duration = .milliseconds(500)
    ) {
        self.petDefinition = petDefinition
        self.interactionCooldown = max(interactionCooldown, .zero)
    }

    var status: MotionSchedulerStatus {
        if let interactionCursor {
            return status(for: interactionCursor, isInteraction: true)
        }

        guard let baseCursor else {
            return .stopped
        }

        return status(for: baseCursor, isInteraction: false)
    }

    var activeSequenceID: String? {
        interactionCursor?.sequence.id ?? baseCursor?.sequence.id
    }

    var pendingSequenceID: String? {
        pendingSequence?.id
    }

    var activeStepRemainingDuration: Duration? {
        interactionCursor?.remainingDuration ?? baseCursor?.remainingDuration
    }

    var isInteractionPlaying: Bool {
        interactionCursor != nil
    }

    @discardableResult
    mutating func request(_ sequence: BehaviorSequence) -> Bool {
        guard Self.isValid(sequence) else {
            return false
        }

        guard var baseCursor else {
            self.baseCursor = Cursor(sequence: sequence)
            pendingSequence = nil
            return true
        }

        if baseCursor.sequence.id == sequence.id {
            pendingSequence = nil
            return false
        }

        if baseCursor.isComplete {
            baseCursor = Cursor(sequence: sequence)
            self.baseCursor = baseCursor
            pendingSequence = nil
            return true
        }

        guard pendingSequence?.id != sequence.id else {
            return false
        }

        pendingSequence = sequence
        return true
    }

    @discardableResult
    mutating func triggerInteraction(_ sequence: BehaviorSequence) -> Bool {
        guard
            interactionCursor == nil,
            cooldownRemaining <= .zero,
            Self.isValid(sequence)
        else {
            return false
        }

        interactionCursor = Cursor(sequence: sequence)
        return true
    }

    mutating func advance(by elapsed: Duration) {
        guard !isPaused, elapsed > .zero else {
            return
        }

        var remainingElapsed = elapsed
        if interactionCursor != nil {
            remainingElapsed = advanceInteraction(by: remainingElapsed)
            guard interactionCursor == nil else {
                return
            }
        }

        reduceCooldown(by: remainingElapsed)
        advanceBase(by: remainingElapsed)
    }

    mutating func pause() {
        isPaused = true
    }

    mutating func resume() {
        isPaused = false
    }

    mutating func stop() {
        baseCursor = nil
        pendingSequence = nil
        interactionCursor = nil
        cooldownRemaining = .zero
        isPaused = false
    }

    private func status(
        for cursor: Cursor,
        isInteraction: Bool
    ) -> MotionSchedulerStatus {
        guard cursor.sequence.steps.indices.contains(cursor.stepIndex) else {
            return .unavailable
        }

        let step = cursor.sequence.steps[cursor.stepIndex]
        guard let motion = petDefinition.motion(id: step.motionID)
            ?? petDefinition.motion(id: "idle")
        else {
            return .unavailable
        }

        return .playing(
            ScheduledMotion(
                sequenceID: cursor.sequence.id,
                stepIndex: cursor.stepIndex,
                requestedMotionID: step.motionID,
                motion: motion,
                playbackSpeed: step.playbackSpeed,
                isInteraction: isInteraction
            )
        )
    }

    private mutating func advanceInteraction(by elapsed: Duration) -> Duration {
        var remainingElapsed = elapsed

        while remainingElapsed > .zero, var cursor = interactionCursor {
            if remainingElapsed < cursor.remainingDuration {
                cursor.remainingDuration -= remainingElapsed
                interactionCursor = cursor
                return .zero
            }

            remainingElapsed -= cursor.remainingDuration
            let nextStepIndex = cursor.stepIndex + 1
            if cursor.sequence.steps.indices.contains(nextStepIndex) {
                cursor.stepIndex = nextStepIndex
                cursor.remainingDuration = cursor.sequence.steps[nextStepIndex].duration
                interactionCursor = cursor
            } else {
                interactionCursor = nil
                cooldownRemaining = interactionCooldown
            }
        }

        return remainingElapsed
    }

    private mutating func advanceBase(by elapsed: Duration) {
        var remainingElapsed = elapsed

        while remainingElapsed > .zero, var cursor = baseCursor, !cursor.isComplete {
            if remainingElapsed < cursor.remainingDuration {
                cursor.remainingDuration -= remainingElapsed
                baseCursor = cursor
                return
            }

            remainingElapsed -= cursor.remainingDuration

            if let pendingSequence {
                baseCursor = Cursor(sequence: pendingSequence)
                self.pendingSequence = nil
                continue
            }

            let nextStepIndex = cursor.stepIndex + 1
            if cursor.sequence.steps.indices.contains(nextStepIndex) {
                cursor.stepIndex = nextStepIndex
                cursor.remainingDuration = cursor.sequence.steps[nextStepIndex].duration
            } else if cursor.sequence.repeats {
                cursor.stepIndex = 0
                cursor.remainingDuration = cursor.sequence.steps[0].duration
            } else {
                cursor.remainingDuration = .zero
                cursor.isComplete = true
            }

            baseCursor = cursor
        }
    }

    private mutating func reduceCooldown(by elapsed: Duration) {
        guard cooldownRemaining > .zero, elapsed > .zero else {
            return
        }

        cooldownRemaining = max(cooldownRemaining - elapsed, .zero)
    }

    private static func isValid(_ sequence: BehaviorSequence) -> Bool {
        !sequence.steps.isEmpty && sequence.steps.allSatisfy {
            !$0.motionID.isEmpty
                && $0.duration > .zero
                && $0.playbackSpeed.isFinite
                && $0.playbackSpeed > 0
        }
    }
}

File: apps/windows/src/MonglePet.Core/Behavior/MotionScheduler.cs
namespace MonglePet.Core.Behavior;

public static class BehaviorMotionReferences
{
    public const string CurrentPetDefault = "__monglepet_current_pet_default__";
    public const string DefaultSequence = "__monglepet_default_behavior__";
}

public sealed record MotionDefinition(string Id, TimeSpan CycleDuration);

public enum MotionSequencePlayback
{
    Stored,
    Once,
    RepeatWhileRequested,
}

public sealed record ScheduledMotion(
    string SequenceId,
    int StepIndex,
    string RequestedMotionId,
    string ResolvedMotionId)
{
    public bool UsesFallback =>
        RequestedMotionId != BehaviorMotionReferences.CurrentPetDefault &&
        !string.Equals(RequestedMotionId, ResolvedMotionId, StringComparison.Ordinal);
}

public abstract record MotionSchedulerStatus
{
    private MotionSchedulerStatus() { }

    public sealed record Stopped : MotionSchedulerStatus;

    public sealed record Playing(ScheduledMotion Motion) : MotionSchedulerStatus;

    public sealed record Completed(string SequenceId) : MotionSchedulerStatus;

    public sealed record Unavailable : MotionSchedulerStatus;
}

public sealed class MotionScheduler
{
    private readonly IReadOnlyDictionary<string, TimeSpan> _cycleDurations;
    private readonly string _defaultMotionId;
    private Cursor? _cursor;
    private bool _unavailable;

    public MotionScheduler(
        string defaultMotionId,
        IReadOnlyDictionary<string, TimeSpan> cycleDurations)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(defaultMotionId);
        ArgumentNullException.ThrowIfNull(cycleDurations);
        if (!cycleDurations.TryGetValue(defaultMotionId, out TimeSpan duration) ||
            duration <= TimeSpan.Zero)
        {
            throw new ArgumentException(
                "The default motion must have a positive cycle duration.",
                nameof(cycleDurations));
        }
        if (cycleDurations.Any(item =>
                string.IsNullOrWhiteSpace(item.Key) || item.Value <= TimeSpan.Zero))
        {
            throw new ArgumentException(
                "Every motion must have an identifier and a positive cycle duration.",
                nameof(cycleDurations));
        }

        _defaultMotionId = defaultMotionId;
        _cycleDurations = cycleDurations;
    }

    public MotionSchedulerStatus Status
    {
        get
        {
            if (_cursor is null)
            {
                return _unavailable
                    ? new MotionSchedulerStatus.Unavailable()
                    : new MotionSchedulerStatus.Stopped();
            }
            if (_cursor.IsComplete)
            {
                return new MotionSchedulerStatus.Completed(_cursor.Sequence.Id);
            }

            return new MotionSchedulerStatus.Playing(Scheduled(_cursor, _cursor.StepIndex));
        }
    }

    public bool IsPaused { get; private set; }

    public string? ActiveSequenceId => _cursor?.Sequence.Id;

    public ulong CompletedSequencePassCount { get; private set; }

    public bool IsSequenceComplete => _cursor?.IsComplete ?? false;

    public ScheduledMotion? CompletedMotion => _cursor is { IsComplete: true } cursor
        ? Scheduled(cursor, cursor.Steps.Count - 1)
        : null;

    public TimeSpan? ActiveCycleRemainingDuration =>
        _cursor is { IsComplete: false } cursor
            ? cursor.RemainingCycleDuration
            : null;

    public TimeSpan? ActiveCycleElapsedDuration =>
        _cursor is { IsComplete: false } cursor
            ? cursor.Steps[cursor.StepIndex].CycleDuration - cursor.RemainingCycleDuration
            : null;

    public bool Request(
        BehaviorSequence sequence,
        MotionSequencePlayback playback = MotionSequencePlayback.Stored)
    {
        ArgumentNullException.ThrowIfNull(sequence);
        sequence = ApplyPlayback(sequence, playback);
        Cursor? requested = MakeCursor(sequence);
        if (requested is null)
        {
            if (_cursor is null)
            {
                _unavailable = true;
            }
            return false;
        }

        if (_cursor is not null && SequencesEqual(_cursor.Sequence, sequence))
        {
            _unavailable = false;
            return false;
        }

        _cursor = requested;
        _unavailable = false;
        return true;
    }

    public bool Restart(
        BehaviorSequence sequence,
        MotionSequencePlayback playback = MotionSequencePlayback.Stored)
    {
        ArgumentNullException.ThrowIfNull(sequence);
        sequence = ApplyPlayback(sequence, playback);
        Cursor? requested = MakeCursor(sequence);
        if (requested is null)
        {
            if (_cursor is null)
            {
                _unavailable = true;
            }
            return false;
        }

        _cursor = requested;
        _unavailable = false;
        return true;
    }

    public void Advance(TimeSpan elapsed)
    {
        if (IsPaused || elapsed <= TimeSpan.Zero || _cursor is null || _cursor.IsComplete)
        {
            return;
        }

        TimeSpan remainingElapsed = elapsed;
        while (remainingElapsed > TimeSpan.Zero && !_cursor.IsComplete)
        {
            if (remainingElapsed < _cursor.RemainingCycleDuration)
            {
                _cursor.RemainingCycleDuration -= remainingElapsed;
                return;
            }

            remainingElapsed -= _cursor.RemainingCycleDuration;
            bool completesSequencePass = CycleCompletesSequencePass(_cursor);
            AdvanceAfterCycle(_cursor);
            if (completesSequencePass)
            {
                unchecked
                {
                    CompletedSequencePassCount++;
                }
            }
        }
    }

    public void Pause() => IsPaused = true;

    public void Resume() => IsPaused = false;

    public void Stop()
    {
        _cursor = null;
        _unavailable = false;
        IsPaused = false;
        CompletedSequencePassCount = 0;
    }

    private static bool CycleCompletesSequencePass(Cursor cursor)
    {
        ResolvedStep step = cursor.Steps[cursor.StepIndex];
        return cursor.StepIndex == cursor.Steps.Count - 1 &&
            cursor.CompletedCycles + 1 >= step.Source.RepeatCount;
    }

    private Cursor? MakeCursor(BehaviorSequence sequence)
    {
        if (string.IsNullOrWhiteSpace(sequence.Id) || sequence.Steps.Count == 0)
        {
            return null;
        }

        var steps = new List<ResolvedStep>(sequence.Steps.Count);
        foreach (BehaviorStep step in sequence.Steps)
        {
            if (string.IsNullOrWhiteSpace(step.MotionId) || step.RepeatCount < 1)
            {
                return null;
            }

            string resolvedMotionId = ResolveMotionId(step.MotionId);
            steps.Add(new ResolvedStep(
                step,
                resolvedMotionId,
                _cycleDurations[resolvedMotionId]));
        }

        return new Cursor(sequence, steps);
    }

    private string ResolveMotionId(string requestedMotionId)
    {
        if (string.Equals(
                requestedMotionId,
                BehaviorMotionReferences.CurrentPetDefault,
                StringComparison.Ordinal))
        {
            return _defaultMotionId;
        }

        return _cycleDurations.ContainsKey(requestedMotionId)
            ? requestedMotionId
            : _defaultMotionId;
    }

    private static void AdvanceAfterCycle(Cursor cursor)
    {
        ResolvedStep step = cursor.Steps[cursor.StepIndex];
        cursor.CompletedCycles++;
        if (cursor.CompletedCycles < step.Source.RepeatCount)
        {
            cursor.RemainingCycleDuration = step.CycleDuration;
            return;
        }

        int nextStepIndex = cursor.StepIndex + 1;
        if (nextStepIndex < cursor.Steps.Count)
        {
            cursor.StepIndex = nextStepIndex;
            cursor.CompletedCycles = 0;
            cursor.RemainingCycleDuration = cursor.Steps[nextStepIndex].CycleDuration;
            return;
        }

        if (!cursor.Sequence.Repeats)
        {
            cursor.RemainingCycleDuration = TimeSpan.Zero;
            cursor.IsComplete = true;
            return;
        }

        cursor.StepIndex = 0;
        cursor.CompletedCycles = 0;
        cursor.RemainingCycleDuration = cursor.Steps[0].CycleDuration;
    }

    private static BehaviorSequence ApplyPlayback(
        BehaviorSequence sequence,
        MotionSequencePlayback playback) => playback switch
    {
        MotionSequencePlayback.Once => sequence with { Repeats = false },
        MotionSequencePlayback.RepeatWhileRequested => sequence with { Repeats = true },
        _ => sequence,
    };

    private static ScheduledMotion Scheduled(Cursor cursor, int stepIndex)
    {
        ResolvedStep step = cursor.Steps[stepIndex];
        return new ScheduledMotion(
            cursor.Sequence.Id,
            stepIndex,
            step.Source.MotionId,
            step.ResolvedMotionId);
    }

    private static bool SequencesEqual(BehaviorSequence left, BehaviorSequence right)
    {
        if (!string.Equals(left.Id, right.Id, StringComparison.Ordinal) ||
            left.Repeats != right.Repeats ||
            left.Steps.Count != right.Steps.Count)
        {
            return false;
        }

        for (int index = 0; index < left.Steps.Count; index++)
        {
            if (left.Steps[index] != right.Steps[index])
            {
                return false;
            }
        }
        return true;
    }

    private sealed record ResolvedStep(
        BehaviorStep Source,
        string ResolvedMotionId,
        TimeSpan CycleDuration);

    private sealed class Cursor
    {
        public Cursor(BehaviorSequence sequence, IReadOnlyList<ResolvedStep> steps)
        {
            Sequence = sequence;
            Steps = steps;
            RemainingCycleDuration = steps[0].CycleDuration;
        }

        public BehaviorSequence Sequence { get; }

        public IReadOnlyList<ResolvedStep> Steps { get; }

        public int StepIndex { get; set; }

        public int CompletedCycles { get; set; }

        public TimeSpan RemainingCycleDuration { get; set; }

        public bool IsComplete { get; set; }
    }
}
